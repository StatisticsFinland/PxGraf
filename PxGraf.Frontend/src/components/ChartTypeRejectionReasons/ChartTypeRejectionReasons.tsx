import { useTranslation } from 'react-i18next';
import React from 'react';
import { MultiLanguageString } from 'types/multiLanguageString';
import {
    Button, Dialog, DialogTitle,
    DialogContent, DialogActions
} from '@mui/material';
import QuestionMarkIcon from '@mui/icons-material/QuestionMark';
import UiLanguageContext from 'contexts/uiLanguageContext';

interface IChartTypeRejectionReasonsProps {
    rejectionReasons: { [key: string]: MultiLanguageString };
}

export const ChartTypeRejectionReasons: React.FC<IChartTypeRejectionReasonsProps> = ({ rejectionReasons }) => {
    const { t } = useTranslation();
    const { language } = React.useContext(UiLanguageContext);
    const [open, setOpen] = React.useState(false);

    const rejectionReasonElements: React.ReactNode[] = Object.keys(rejectionReasons).map(key => {
        return (
            <div key={key}>
                <span>{`${t('chartTypes.' + key)}:`}</span>
                <ul>
                    <li>
                        {rejectionReasons[key][language]}
                    </li>
                </ul>
            </div>
        );
    });
    if (!rejectionReasons) {
        return <></>;
    }
    return (
        <>
            <Button aria-haspopup={true} aria-controls='rejection-dialog' variant={'outlined'} startIcon={<QuestionMarkIcon />} onClick={() => setOpen(true)}>{t("rejectionDialog.buttonText")}</Button>
            <Dialog
                open={open}
                onClose={() => setOpen(false)}
                scroll='paper'
                aria-labelledby="scroll-dialog-title"
                aria-describedby="scroll-dialog-description"
            >
                <DialogTitle id="scroll-dialog-title">{t("rejectionDialog.titleText")}</DialogTitle>
                <DialogContent dividers={true} id="scroll-dialog-description">
                    {rejectionReasonElements}
                </DialogContent>
                <DialogActions>
                    <Button onClick={() => setOpen(false)}>{t("rejectionDialog.closeButton")}</Button>
                </DialogActions>
            </Dialog>
        </>
    );
}

export default ChartTypeRejectionReasons;