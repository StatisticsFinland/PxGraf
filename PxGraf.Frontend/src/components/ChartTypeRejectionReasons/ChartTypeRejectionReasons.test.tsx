import { fireEvent, render, screen } from '@testing-library/react';
import UiLanguageContext from 'contexts/uiLanguageContext';
import React from 'react';
import ChartTypeRejectionReasons from './ChartTypeRejectionReasons';

jest.mock('react-i18next', () => ({
    ...jest.requireActual('react-i18next'),
    useTranslation: () => {
        return {
            t: (str: string) => str,
            i18n: {
                changeLanguage: () => new Promise(() => null),
            },
        };
    },
}));

const setLanguage = jest.fn();
const language = 'fi';
const setLanguageTab = jest.fn();
const languageTab = 'fi';
const availableUiLanguages = ['fi', 'en', 'sv'];
const uiContentLanguage = 'fi';
const setUiContentLanguage = jest.fn();

const mockRejectionReasons = {
    "LineChart": {
        'fi': 'huono kuvio'
    },
    "PieChart": {
        'fi': 'vielÃ¤ huonompi kuvio'
    }
}

describe('Rendering test', () => {
    it('renders correctly with values within range', () => {
        const { asFragment } = render(<UiLanguageContext.Provider value={{ language, setLanguage, languageTab, setLanguageTab, availableUiLanguages, uiContentLanguage, setUiContentLanguage }}><ChartTypeRejectionReasons rejectionReasons={mockRejectionReasons} /></UiLanguageContext.Provider>);
        expect(asFragment()).toMatchSnapshot();
    });
});

describe('Assertion test', () => {
    it('should open the dialog when the button is clicked', () => {
        render(<UiLanguageContext.Provider value={{ language, setLanguage, languageTab, setLanguageTab, availableUiLanguages, uiContentLanguage, setUiContentLanguage }}><ChartTypeRejectionReasons rejectionReasons={mockRejectionReasons} /></UiLanguageContext.Provider>);
        fireEvent.click(screen.getByText('rejectionDialog.buttonText'));
        expect(screen.getByText(mockRejectionReasons['LineChart']['fi'])).toBeInTheDocument();
        expect(screen.getByText(mockRejectionReasons['PieChart']['fi'])).toBeInTheDocument();
    });
});