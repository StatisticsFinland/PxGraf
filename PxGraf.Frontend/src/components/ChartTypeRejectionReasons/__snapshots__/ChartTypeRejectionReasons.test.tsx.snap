// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering test renders correctly with values within range 1`] = `
<DocumentFragment>
  <button
    aria-controls="rejection-dialog"
    aria-haspopup="true"
    class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary css-1rwt2y5-MuiButtonBase-root-MuiButton-root"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiButton-icon MuiButton-startIcon MuiButton-iconSizeMedium css-1d6wzja-MuiButton-startIcon"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="QuestionMarkIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M11.07 12.85c.77-1.39 2.25-2.21 3.11-3.44.91-1.29.4-3.7-2.18-3.7-1.69 0-2.52 1.28-2.87 2.34L6.54 6.96C7.25 4.83 9.18 3 11.99 3c2.35 0 3.96 1.07 4.78 2.41.7 1.15 1.11 3.3.03 4.9-1.2 1.77-2.35 2.31-2.97 3.45-.25.46-.35.76-.35 2.24h-2.89c-.01-.78-.13-2.05.48-3.15M14 20c0 1.1-.9 2-2 2s-2-.9-2-2 .9-2 2-2 2 .9 2 2"
        />
      </svg>
    </span>
    rejectionDialog.buttonText
    <span
      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </button>
</DocumentFragment>
`;
